kernel/src/scan/data_skipping.rs

```diff
@@ -104,62 +60,89 @@ fn as_inverted_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
 ///         are not eligible for data skipping.
 /// - `OR` is rewritten only if all operands are eligible for data skipping. Otherwise, the whole OR
 ///        expression is dropped.
-fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
+fn as_data_skipping_predicate(expr: &Expr, inverted: bool) -> Option<Expr> {
     use BinaryOperator::*;
     use Expr::*;
     use UnaryOperator::*;
 
     match expr {
+        Literal(Scalar::Boolean(value)) => Some(Expr::literal(*value != inverted)),
+        Literal(_) => None, // Non-boolean literals unsupported
+        // The expression <col> is equivalent to <col> != FALSE. Resembling NotEqual case below.
+        Column(col) => Some(Expr::or(
```

Comment:
This causes an error if the column isn't a boolean column, even if it could be coerced to a boolean.

So a query that might work:

```sql
SELECT * FROM foo WHERE int_col
```

will cause kernel to error.

Not sure what the best course of action is wrt. to doing coercion in our expression eval, checking if the col is bool here, or just giving up on col expressions for now.

This is actually why the test for expression:
```rust
            Expression::column("number").and(empty_struct.clone().is_null()),
```

is failing.

Thoughts?

Reply:
Expression framework docs are very clear that kernel always expects exact type matches, and that engine is responsible to apply any casting or coercion that should happen. Granted, we don't currently provide a cast expression, so that could be difficult to do in practice. 

I guess the real question is -- is kernel responsible to prevent/swallow all query errors that might result from a data skipping predicate that was derived from a syntactically incorrect WHERE clause the engine gave it?

Reply:
So we would expect that if the engine sees the user has written `WHERE int_col` it would rewrite that as `WHERE (cast int_col as boolean)` and then give that to kernel (which as you mention is hard given we don't have a cast)?